snakes_battle/logic.py

```diff
@@ -50,20 +63,64 @@ def apply_logic(board):
         for _snake in board.snakes:
 
             if snake == _snake: # snake hitted itself.
-                if snake.body_pos[0] in _snake.body_pos[1:]:
-                    snake_lost(snake,board)
-                    break
 
-            elif snake.body_pos[0] in _snake.body_pos: # snake hitted other snakes
-                snake_lost(snake,board)
-                break
+                headless = _snake.body_pos[1:] # Snake body position without the head
+
+                for i in range(len(headless)):
+                    if snake.body_pos[0] == headless[i]:
+                        if snake.super_power["SHIELD"]: # If snake is shielded then it won't shrink
+                            snake.super_power["SHIELD"] = False
+                        else:
+                            snake.shrink(len(headless) - i) # Snake is not shielded so it will be shrinked. Snake cuts itself.
+                            break
+
+            else: # snake hitted other snakes
+
+                for i in range(len(_snake.body_pos) - 1):
+
+                    if snake.body_pos[0] == _snake.body_pos[i]:
+
+                        if snake.super_power["KNIFE"]:
+                            snake.super_power["KNIFE"] = False
+                            _snake.shrink(len(_snake.body_pos) - i)
+                            break
+
+                        elif snake.super_power["SHIELD"]: # Snake can hit other snakes without lose if it shielded
+                                snake.super_power["SHIELD"] = False
+                        
+                        else: # Snake not shielded so it loses.
+                            snake_lost(snake,board)
+                            break
+            
         
     # Rule - generate a bomb randomly.
-    if (random.random() < settings.BOMB_CREATION_PROBABILITY):
+    if (random.random() < FruitKind.BOMB["creation_probability"]):
         new_bomb = Fruit(FruitKind.BOMB, get_new_fruit_position(board))
         board.add_fruit(new_bomb)
+    
+    # Rule - generate a special fruit randomly.
+    if random.random() < FruitKind.SHIELD["creation_probability"]:
```

Comment:
There's a better way to do that if you run over a list containing referenced to all fruit kinds.
Just code duplicated for every fruit, no real reason.